### PR TITLE
Allow custom template function

### PR DESCRIPTION
**Pull request by davidsoderberg**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Check if template is a function and then use result from function as content in handler.

- **What is the current behavior?** (You can also link to an open issue here)
Only text and handlebars templates are allowed.

- **What is the new behavior (if this is a feature change)?**
Check if template is a function and then use result from function as content in handler.
https://github.com/notifirehq/notifire/issues/31

- **Other information**:
This is one of my first PR against an open source project 😄

### CODE REVIEW

File: packages/core/src/lib/handler/sms.handler.spec.ts [hunk 1]
```diff
@@ -28,8 +29,36 @@ test('send sms should call the provider method correctly', async () => {
...
+    firstName: 'test name',
+  });
+
+  expect(spyTemplateFunction).toHaveBeenCalled();
+  expect(spyTemplateFunction).toBeCalledWith({
+    $email: 'test@email.com',
+    $phone: '+1333322214',
+    $user_id: '1234',
+    firstName: 'test name',
+  });
```
Comment (scopsy) [inline packages/core/src/lib/handler/sms.handler.spec.ts:63]: @davidsoderberg Maybe it's better to create a new `test` case to specifically test the template function?

File: packages/core/src/lib/handler/sms.handler.spec.ts [hunk 2]
```diff
@@ -28,8 +29,36 @@ test('send sms should call the provider method correctly', async () => {
...
+    content: 'Name: test name',
+    to: '+1333322214',
   });
   spy.mockRestore();
+
+  const spyTemplateFunction = jest
+    .fn()
+    .mockImplementation(() => Promise.resolve('test'));
+
+  emailHandler = new SmsHandler(
```
Comment (scopsy) [inline packages/core/src/lib/handler/sms.handler.spec.ts:41]: I believe it should be smsHandler :) I now see that the mistake is also in the code above.

File: packages/core/src/lib/handler/sms.handler.spec.ts [hunk 3]
```diff
@@ -6,11 +6,12 @@ test('send sms should call the provider method correctly', async () => {
...
     id: 'sms-provider',
     channelType: ChannelTypeEnum.SMS,
-    sendMessage: () => Promise.resolve({ id: '1', date: new Date().toString() }),
+    sendMessage: () =>
+      Promise.resolve({ id: '1', date: new Date().toString() }),
   };
 
   const spy = jest.spyOn(provider, 'sendMessage');
-  const emailHandler = new SmsHandler(
+  let emailHandler = new SmsHandler(
```
Comment (scopsy) [inline packages/core/src/lib/handler/sms.handler.spec.ts:14]: Extracting the bottom test should also keep this test with `const`

File: packages/core/src/lib/handler/email.handler.spec.ts [hunk 4]
```diff
@@ -41,17 +42,44 @@ test('send should call the provider method correctly', async () => {
...
+    firstName: 'test name',
+  });
+
+  expect(spyTemplateFunction).toHaveBeenCalled();
+  expect(spyTemplateFunction).toBeCalledWith({
+    $branding: {},
+    $email: 'test@email.com',
+    $user_id: '1234',
+    firstName: 'test name',
+  });
```
Comment (scopsy) [inline packages/core/src/lib/handler/email.handler.spec.ts:72]: Same here, suggest extracting this to a separate test case

File: packages/core/src/lib/trigger/trigger.engine.spec.ts [hunk 5]
```diff
@@ -88,4 +90,22 @@ test('variable protection should throw if missing variable provided', async () =
...
+      },
+    ],
+  });
+
+  await expect(
+    triggerEngine.trigger('test-notification-promise', {
+      $user_id: '12345',
+      $email: 'test@gmail.com',
+    })
+  ).rejects.toEqual(new Error('Missing variables passed. firstName'));
```
Comment (scopsy) [inline packages/core/src/lib/trigger/trigger.engine.spec.ts:110]: I would create a new test case for this

**Review by scopsy — COMMENTED** [review]
Looks amazing @davidsoderberg !! Added some comments mainly about the structure of the test.

**Comment by scopsy** [comment]
And congrats on one of your first PR's! This is incredible work 🎉🚀

**Comment by davidsoderberg** [comment]
Thanks, I think I managed to fix according to your comments 😄

**Review by scopsy — APPROVED** [review]

**Pull request merged**